File: ios/Classes/QYPlatformView.m
//
//  QYPlatformView.m
//  qy_player
//
//  Created by 捡来的💻 on 2022/6/27.
//

#import "QYPlatformView.h"
#import <qysdk/qysdk.h>


@implementation QYPlatformViewFactory{
    NSObject<FlutterBinaryMessenger>* _messenger;
  }

- (instancetype)initWithMessenger:(NSObject<FlutterBinaryMessenger>*)messenger {
  self = [super init];
  if (self) {
    _messenger = messenger;
  }
  return self;
}

- (NSObject<FlutterPlatformView>*)createWithFrame:(CGRect)frame
                                   viewIdentifier:(int64_t)viewId
                                        arguments:(id _Nullable)args {
  return [[QYPlatformView alloc] initWithFrame:frame
                              viewIdentifier:viewId
                                   arguments:args
                             binaryMessenger:_messenger];
}

@end

@interface QYPlatformView ()<QYPlayerDelegate>

@end

@implementation QYPlatformView {
    UIView *view;
    NSObject<FlutterBinaryMessenger>* _messenger;
    QYSession *session;
    QYPlayer *player;
    QYPlayer *talkPlayer;
}


- (instancetype)initWithFrame:(CGRect)frame
               viewIdentifier:(int64_t)viewId
                    arguments:(id _Nullable)args
              binaryMessenger:(NSObject<FlutterBinaryMessenger>*)messenger {
  if (self = [super init]) {
      view = [[UIView alloc] init];
      session = [[QYSession alloc] init];
      _messenger = messenger;
      [self initMethodChannel];
  }
  return self;
}


- (void)initMethodChannel {
    NSLog(@"初始化消息通道");
    FlutterMethodChannel *channel = [FlutterMethodChannel methodChannelWithName:@"qy_player" binaryMessenger:_messenger];
    [channel setMethodCallHandler:^(FlutterMethodCall * _Nonnull call, FlutterResult  _Nonnull result) {
        NSLog(@"接收到消息 --- [%@]",call.method);
        if ([@"preview" isEqualToString:call.method]) {
            NSDictionary *data = call.arguments;
            NSString *viewUrl = data[@"viewUrl"];
            NSLog(@"preview --- [%@]",viewUrl);
            if (self->player != nil) {
                [self->player Release];
                self->player = nil;
            }
            self->player = [[QYPlayer alloc] init];
            [self->player SetEventDelegate:self];
            [self->player SetCanvas: self->view];
            [self->player CtrlAudio:YES];
            NSLog(@"player = %@",self->player);
//            [self->player SyncConnect:viewUrl];
            [self->player Connect:viewUrl CallBack:^(int32_t ret) {
                if(ret==0){;
                    NSLog(@"预览成功");
                }else{
                    NSLog(@"预览失败");
                }
            }];
            result(@"开始播放");
        } else  if ([@"stop_preview" isEqualToString:call.method]) {
            [self->player Release];
            self->player = nil;
            result(@"开始播放");
        }   else  if ([@"ctrolSound" isEqualToString:call.method]) {
            NSDictionary *data = call.arguments;
            bool soud = [data[@"isOpen"] boolValue];
            [self->player CtrlAudio:soud];
            result(@"开始播放");
        } else if ([@"talk" isEqualToString:call.method]) {
            NSDictionary *data = call.arguments;
            NSString *viewUrl = data[@"viewUrl"];
            NSLog(@"Talk --- [%@]",viewUrl);
            if (self->talkPlayer != nil) {
                [self->talkPlayer Release];
                self->talkPlayer = nil;
            }
            self->talkPlayer = [[QYPlayer alloc] init];
            [self->talkPlayer SetEventDelegate:self];
            [self->talkPlayer SetCanvas: self->view];
            [self->talkPlayer CtrlAudio:YES];
            NSLog(@"player = %@",self->talkPlayer);
//            [self->player SyncConnect:viewUrl];
            [self->talkPlayer Connect:viewUrl CallBack:^(int32_t ret) {
                if(ret==0){;
                    NSLog(@"开启对讲成功");
                    [self->talkPlayer CtrlTalk:true];
                }else{
                    NSLog(@"预览失败");
                }
            }];
            result(@"开始播放");
        } else  if ([@"stop_talk" isEqualToString:call.method]) {
            [self->talkPlayer CtrlTalk:NO];
            [self->talkPlayer Release];
            self->talkPlayer = nil;
            result(@"开始播放");
        }
        else {
            result(FlutterMethodNotImplemented);
        }
    }];
}




#pragma mark - PlayerDelegate Methods 播放器回调
-(void)onAudioCallBack:(int) type
              AudioType:(int) audiotype
                 Smaple:(int) smpale
                    Chn:(int) chn
                    Bit:(int) bit
                   Data:(NSData*)  data
                 Length:(int)  len
{
    
    NSLog(@"-------");
    
}



-(void)onVideoCallBack:(int) type
                 Iframe:(int) iframe
              TimeStamp:(long long) timestamp
                   Data:(NSData*)  data
                 Length:(int)  len
{
    
    NSLog(@"-------");
}

-(void)onDisConnect
{
    
    NSLog(@"-------");
}

- (UIView*)view {
  return view;
}

@end

File: android/src/main/java/com/ttxs/qy_player/QYPlatformView.java
package com.ttxs.qy_player;

import android.app.Application;
import android.content.Context;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.wholeally.qysdk.QyPlayer;
import com.wholeally.qysdk.QySession;
import com.wholeally.qysdk.event.QyPlayerEvent;
import com.wholeally.qysdk.implement.QyPlayerImplement;
import com.wholeally.qysdk.implement.QySessionImplement;
import com.wholeally.yuv.GLFrameRenderer;
import com.wholeally.yuv.GLFrameSurface;

import java.util.Map;

import io.flutter.Log;
import io.flutter.plugin.common.BasicMessageChannel;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.StandardMessageCodec;
import io.flutter.plugin.platform.PlatformView;

public class QYPlatformView implements PlatformView, MethodChannel.MethodCallHandler {
    @NonNull
    private final BinaryMessenger messenger;
    private static final String CHANNEL = "qy_player";
    private static final String PLUGIN_TAG = "QYPlugin";
    private Application application;
    private GLFrameSurface surfaceView;
    private GLFrameRenderer glFrameRenderer;
    private QySession session;
    private QyPlayer player;
    private QyPlayer talker;
    BasicMessageChannel<Object> messageChannel;

    public QYPlatformView(@NonNull Context context, int id, @Nullable Map<String, Object> creationParams, BinaryMessenger messenger, Application application) {
        this.messenger = messenger;
//        this.context = context;
        this.application = application;
        new MethodChannel(messenger, CHANNEL).setMethodCallHandler(this);
        messageChannel = new BasicMessageChannel<>(messenger, "qy_player/message", new StandardMessageCodec());


        surfaceView = new GLFrameSurface(context);
        surfaceView.setEGLContextClientVersion(2);
        glFrameRenderer = new GLFrameRenderer(surfaceView, context);
        surfaceView.setRenderer(glFrameRenderer);

        session = new QySessionImplement(application);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
         if (call.method.equals("preview")) {
            /// 通过URL预览
            String viewUrl = call.argument("viewUrl");
            previewWithUrl(viewUrl);
        } else if (call.method.equals("stop_preview")) {
             if (player != null) {
                 player.Close();
                 player = null;
             }
            result.success("success");
        } else if (call.method.equals("talk")) {
            /// 通过URL预览
            String viewUrl = call.argument("viewUrl");
            talkWithUrl(viewUrl);
        } else if (call.method.equals("stop_talk")) {

             if (talker != null) {
                 talker.Talk(false);
                 talker.Close();
                 talker = null;
             }
            result.success("success");
        } else if (call.method.equals("ctrolSound")) {
            Boolean soud = call.argument("isOpen");
            if (soud) {
                player.CtrlAudio(true);
            } else {
                player.CtrlAudio(false);
            }
            result.success("success");
        } else {
            result.notImplemented();
        }
    }



    public void previewWithUrl(String viewUrl) {
        Log.d(PLUGIN_TAG,"开始预览 :" + viewUrl);
        player = new QyPlayerImplement(application);
        player.SetCanvas(glFrameRenderer);
        player.SetPlayerEvent(new QyPlayerEvent() {
            @Override
            public void onVideoDataCallBack(int videoType, int iframe, long timeStamp,
                                            final byte[] data, int dataLength) {
                System.out.println("===wholeally_111_onVideoDataCallBack_iframe=======:"+iframe
                        +";=timeStamp=:"+timeStamp+";=videoType=:"+videoType);
            }
            @Override
            public void onYuvDataCallBack(int iframe, long timeStamp, byte[] data, int
                    dataLength, int width, int height) {
                System.out.println("===wholeally_111_onYuvDataCallBack===:"+width+
                        ";=height=:"+height);
//                                                Log.e(TAG_PLAYER, "_____Yuv数据____:" + iframe + ";;" + timeStamp + ";;" +
//                                                        "" + data + ";;" + dataLength + ";;" + width + ";;" + height);
            }
            @Override
            public void onAudioDataCallBack(int type, int audiotype, int smaple, int
                    chn, int bit, byte[] data, int dataLength) {
                Log.e(PLUGIN_TAG, "_____Audio数据____:" + type + ";;" + audiotype + ";;" +
                        "" + smaple + ";;" + chn + ";;" + bit + ";;" + data + ";;" +
                        dataLength);
            }
            @Override
            public void onDisConnect() {
//                                                Log.e(TAG_PLAYER, "_____断开通知____");
            }
        });
        player.Connect(viewUrl, new QyPlayer
                .OnCommonCallBack() {
            @Override
            public void on(int ret) {
                if (ret == 0) {
                    Log.d(PLUGIN_TAG,  "预览成功");
                } else {
                    Log.d(PLUGIN_TAG,  "预览失败:" );
                }
            }
        });
    }

    public void talkWithUrl(String viewUrl) {
        Log.d(PLUGIN_TAG,"开始对讲 :" + viewUrl);
        talker = new QyPlayerImplement(application);
        talker.SetCanvas(glFrameRenderer);
        talker.SetPlayerEvent(new QyPlayerEvent() {
            @Override
            public void onVideoDataCallBack(int videoType, int iframe, long timeStamp,
                                            final byte[] data, int dataLength) {
                System.out.println("===wholeally_111_onVideoDataCallBack_iframe=======:"+iframe
                        +";=timeStamp=:"+timeStamp+";=videoType=:"+videoType);
            }
            @Override
            public void onYuvDataCallBack(int iframe, long timeStamp, byte[] data, int
                    dataLength, int width, int height) {
                System.out.println("===wholeally_111_onYuvDataCallBack===:"+width+
                        ";=height=:"+height);
//                                                Log.e(TAG_PLAYER, "_____Yuv数据____:" + iframe + ";;" + timeStamp + ";;" +
//                                                        "" + data + ";;" + dataLength + ";;" + width + ";;" + height);
            }
            @Override
            public void onAudioDataCallBack(int type, int audiotype, int smaple, int
                    chn, int bit, byte[] data, int dataLength) {
                Log.e(PLUGIN_TAG, "_____Audio数据____:" + type + ";;" + audiotype + ";;" +
                        "" + smaple + ";;" + chn + ";;" + bit + ";;" + data + ";;" +
                        dataLength);
            }
            @Override
            public void onDisConnect() {
//                                                Log.e(TAG_PLAYER, "_____断开通知____");
            }
        });
        talker.Connect(viewUrl, new QyPlayer
                .OnCommonCallBack() {
            @Override
            public void on(int ret) {
                if (ret == 0) {
                    talker.Talk(true);
                    Log.d(PLUGIN_TAG,  "预览成功");
                } else {
                    Log.d(PLUGIN_TAG,  "预览失败:" );
                }
            }
        });
    }


    @Override
    public View getView() {
        return surfaceView;
    }

    @Override
    public void dispose() {
        if (null != player) {
            player.Close();
        }
        if (null != talker) {
            talker.Close();
        }
    }
}
